google-ads-examples/src/main/java/com/google/ads/googleads/examples/remarketing/AddCustomerMatchUserList.java

```diff
@@ -312,6 +291,85 @@ private void addUsersToCustomerMatchUserList(
+   *
+   * @param digest the digest to use to hash the normalized string.
+   * @param s the string to normalize and hash.
+   */
+  private String normalizeAndHash(MessageDigest digest, String s)
+      throws UnsupportedEncodingException {
+    // Normalizes by removing leading and trailing whitespace and converting all characters to
+    // lower case.
+    String normalized = s.trim().toLowerCase();
+    // Hashes the normalized string using the hashing algorithm.
+    byte[] hash = digest.digest(normalized.getBytes("UTF-8"));
+    StringBuilder result = new StringBuilder();
+    for (byte b : hash) {
+      result.append(String.format("%02x", b));
+    }
+
+    return result.toString();
+  }
+
+  /**
+   * Retrieves, checks, and prints the status of the offline user data job.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param offlineUserDataJobResourceName the resource name of the OfflineUserDataJob to get the
+   *     status for.
+   * @param userListResourceName the resource name of the Customer Match user list.
+   */
+  private void checkJobStatus(
+      GoogleAdsClient googleAdsClient,
+      long customerId,
+      String offlineUserDataJobResourceName,
+      String userListResourceName) {
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      String query =
+          String.format(
+              "SELECT offline_user_data_job.resource_name, "
+                  + "offline_user_data_job.id, "
+                  + "offline_user_data_job.status, "
+                  + "offline_user_data_job.type, "
+                  + "offline_user_data_job.failure_reason "
+                  + "FROM offline_user_data_job "
+                  + "WHERE offline_user_data_job.resource_name = '%s'",
+              offlineUserDataJobResourceName);
+      // Issues the query and gets the GoogleAdsRow containing the job from the response.
+      GoogleAdsRow googleAdsRow =
+          googleAdsServiceClient
+              .search(Long.toString(customerId), query)
+              .iterateAll()
+              .iterator()
+              .next();
+      OfflineUserDataJob offlineUserDataJob = googleAdsRow.getOfflineUserDataJob();
+      System.out.printf(
+          "Offline user data job ID %d with type '%s' has status: %s%n",
+          offlineUserDataJob.getId().getValue(),
+          offlineUserDataJob.getType(),
+          offlineUserDataJob.getStatus());
+      OfflineUserDataJobStatus jobStatus = offlineUserDataJob.getStatus();
+      if (OfflineUserDataJobStatus.SUCCESS == jobStatus) {
+        // Prints information about the user list.
+        printCustomerMatchUserListInfo(googleAdsClient, customerId, userListResourceName);
+      } else if (OfflineUserDataJobStatus.FAILED == jobStatus) {
+        System.out.printf("  Failure reason: %s%n", offlineUserDataJob.getFailureReason());
```

Comment:
Is there a good way to get more details about what failed? E.g. can we show how to get the failed operations?

Reply:
If the adding of operations to the job fails, there is a partial failure above that will give more specific information. This failure represents an error in the processing of that job, and I believe that this is as detailed a failure as we can provide.

Reply:
@devchas is correct. By design, the job does not report whether each member operation succeeded or failed.

Reply:
Ah yes, makes sense.